File: backend/app/services/vcard.py
from __future__ import annotations

import re
from typing import Optional


def _escape(value: str) -> str:
    return value.replace("\\", "\\\\").replace(";", "\\;").replace("\n", "\\n")


def build_vcard(
    *,
    full_name: str,
    job_title: Optional[str] = None,
    company: Optional[str] = None,
    phone: Optional[str] = None,
    email: Optional[str] = None,
    website: Optional[str] = None,
    linkedin: Optional[str] = None,
) -> str:
    lines = [
        "BEGIN:VCARD",
        "VERSION:3.0",
        f"FN:{_escape(full_name)}",
        f"N:{_escape(full_name)};;;;",
    ]

    if job_title:
        lines.append(f"TITLE:{_escape(job_title)}")
    if company:
        lines.append(f"ORG:{_escape(company)}")
    if phone:
        digits = re.sub(r"\s", "", phone)
        lines.append(f"TEL;TYPE=CELL:{digits}")
    if email:
        lines.append(f"EMAIL;TYPE=INTERNET:{_escape(email)}")
    if website:
        url = website if website.startswith("http") else f"https://{website}"
        lines.append(f"URL:{url}")
    if linkedin:
        url = linkedin if linkedin.startswith("http") else f"https://linkedin.com/in/{linkedin.lstrip('/')}"
        lines.append(f"URL;TYPE=LinkedIn:{url}")

    lines.append("END:VCARD")
    return "\r\n".join(lines)


def vcard_filename(full_name: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", full_name.lower()).strip("-") or "contact"
    return f"{slug}.vcf"

File: frontend/src/lib/vcard.ts
export type VCardData = {
  fullName: string;
  jobTitle?: string;
  company?: string;
  phone?: string;
  email?: string;
  website?: string;
  linkedin?: string;
};

function escapeVCard(value: string) {
  return value.replace(/\\/g, "\\\\").replace(/;/g, "\\;").replace(/\n/g, "\\n");
}

export function buildVCard(data: VCardData): string {
  const lines = [
    "BEGIN:VCARD",
    "VERSION:3.0",
    `FN:${escapeVCard(data.fullName)}`,
    `N:${escapeVCard(data.fullName)};;;;`,
  ];

  if (data.jobTitle) lines.push(`TITLE:${escapeVCard(data.jobTitle)}`);
  if (data.company) lines.push(`ORG:${escapeVCard(data.company)}`);
  if (data.phone) lines.push(`TEL;TYPE=CELL:${data.phone.replace(/\s/g, "")}`);
  if (data.email) lines.push(`EMAIL;TYPE=INTERNET:${escapeVCard(data.email)}`);
  if (data.website) {
    const url = data.website.startsWith("http") ? data.website : `https://${data.website}`;
    lines.push(`URL:${url}`);
  }
  if (data.linkedin) lines.push(`URL;TYPE=LinkedIn:${data.linkedin}`);

  lines.push("END:VCARD");
  return lines.join("\r\n");
}

export async function saveContact(data: VCardData): Promise<void> {
  const content = buildVCard(data);
  const filename = vcardFilename(data);
  const blob = new Blob([content], { type: "text/vcard;charset=utf-8" });
  const file = new File([content], filename, { type: "text/vcard" });

  if (typeof navigator !== "undefined" && navigator.share) {
    try {
      if (navigator.canShare?.({ files: [file] })) {
        await navigator.share({ files: [file], title: data.fullName });
        return;
      }
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") return;
    }
  }

  const isMobile = typeof navigator !== "undefined" && /iPhone|iPad|iPod|Android/i.test(navigator.userAgent);
  if (isMobile) {
    window.location.assign(`data:text/vcard;charset=utf-8,${encodeURIComponent(content)}`);
    return;
  }

  downloadVCard(data, filename);
}

function vcardFilename(data: VCardData) {
  return `${data.fullName.replace(/\s+/g, "-").toLowerCase() || "contact"}.vcf`;
}

export function downloadVCard(data: VCardData, filename?: string) {
  const content = buildVCard(data);
  const blob = new Blob([content], { type: "text/vcard;charset=utf-8" });
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename ?? `${data.fullName.replace(/\s+/g, "-").toLowerCase() || "contact"}.vcf`;
  link.click();
  URL.revokeObjectURL(url);
}

export function whatsappHref(number: string, message?: string) {
  const digits = number.replace(/\D/g, "");
  const base = `https://wa.me/${digits}`;
  if (!message) return base;
  return `${base}?text=${encodeURIComponent(message)}`;
}

export function normalizeLinkedInUrl(url: string) {
  const trimmed = url.trim();
  if (!trimmed) return "";
  if (trimmed.startsWith("http")) return trimmed;
  if (trimmed.startsWith("linkedin.com") || trimmed.startsWith("www.linkedin.com")) {
    return `https://${trimmed}`;
  }
  return `https://linkedin.com/in/${trimmed.replace(/^\/+/, "")}`;
}
